In [ ]:
from google.colab import drive
drive.mount('/content/drive/', force_remount=True)

Mounted at /content/drive/


In [ ]:
ls

drive/  sample_data/


In [ ]:
cd drive/MyDrive/ObjectDetection-master/

/content/drive/MyDrive/ObjectDetection-master


In [ ]:
ls

annotation.txt           out.csv     README.md  simpsons_dataset/
kaggle_simpson_testset/  output/     save/      train.py
model/                   predict.py  Simpsons/


In [ ]:
!pip install keras==2.2.0

  Using cached https://files.pythonhosted.org/packages/68/12/4cabc5c01451eb3b413d19ea151f36e33026fc0efb932bf51bcaf54acbf5/Keras-2.2.0-py2.py3-none-any.whl
  Using cached https://files.pythonhosted.org/packages/f8/33/275506afe1d96b221f66f95adba94d1b73f6b6087cfb6132a5655b6fe338/Keras_Preprocessing-1.0.1-py2.py3-none-any.whl
  Using cached https://files.pythonhosted.org/packages/e2/60/c557075e586e968d7a9c314aa38c236b37cb3ee6b37e8d57152b1a5e0b47/Keras_Applications-1.0.2-py2.py3-none-any.whl
ERROR: tensorflow 2.4.1 has requirement keras-preprocessing~=1.1.2, but you'll have keras-preprocessing 1.0.1 which is incompatible.
  Found existing installation: Keras-Preprocessing 1.1.2
    Uninstalling Keras-Preprocessing-1.1.2:
      Successfully uninstalled Keras-Preprocessing-1.1.2
  Found existing installation: Keras 2.4.3
    Uninstalling Keras-2.4.3:
      Successfully uninstalled Keras-2.4.3


In [ ]:
!pip install tensorflow==1.15

     |████████████████████████████████| 412.3MB 28kB/s 
     |████████████████████████████████| 3.8MB 40.3MB/s 
     |████████████████████████████████| 51kB 5.7MB/s 
     |████████████████████████████████| 51kB 5.8MB/s 
     |████████████████████████████████| 512kB 40.5MB/s 
  Created wheel for gast: filename=gast-0.2.2-cp37-none-any.whl size=7540 sha256=59c31772991ec447437c7f5cc0e0232f9d0ae11fb940a8286a06b70bb32f5f79
  Stored in directory: /root/.cache/pip/wheels/5c/2e/7e/a1d4d4fcebe6c381f378ce7743a3ced3699feb89bcfbdadadd
Successfully built gast
ERROR: tensorflow-probability 0.12.1 has requirement gast>=0.3.2, but you'll have gast 0.2.2 which is incompatible.
ERROR: keras 2.2.0 has requirement keras-applications==1.0.2, but you'll have keras-applications 1.0.8 which is incompatible.
ERROR: keras 2.2.0 has requirement keras-preprocessing==1.0.1, but you'll have keras-preprocessing 1.1.2 which is incompatible.
  Found existing installation: tensorboard 2.4.1
    Uninstalling tensorboard

In [ ]:
!pip install Keras-Applications==1.0.7

     |████████████████████████████████| 61kB 5.6MB/s 
ERROR: tensorflow 1.15.0 has requirement keras-applications>=1.0.8, but you'll have keras-applications 1.0.7 which is incompatible.
ERROR: keras 2.2.0 has requirement keras-applications==1.0.2, but you'll have keras-applications 1.0.7 which is incompatible.
ERROR: keras 2.2.0 has requirement keras-preprocessing==1.0.1, but you'll have keras-preprocessing 1.1.2 which is incompatible.
  Found existing installation: Keras-Applications 1.0.8
    Uninstalling Keras-Applications-1.0.8:
      Successfully uninstalled Keras-Applications-1.0.8


#[Problem 1] Learning and Estimation

In [ ]:
!python train.py -p annotation.txt

Using TensorFlow backend.
Parsing annotation files
Training images per class (19 classes) :
{'abraham_grampa_simpson': 687,
 'apu_nahasapeemapetilon': 206,
 'bart_simpson': 650,
 'bg': 0,
 'charles_montgomery_burns': 650,
 'chief_wiggum': 209,
 'comic_book_guy': 208,
 'edna_krabappel': 212,
 'homer_simpson': 718,
 'kent_brockman': 213,
 'krusty_the_clown': 429,
 'lisa_simpson': 756,
 'marge_simpson': 629,
 'milhouse_van_houten': 210,
 'moe_szyslak': 403,
 'ned_flanders': 675,
 'nelson_muntz': 219,
 'principal_skinner': 614,
 'sideshow_bob': 201}
-------------------------------
path to config file : ./save/train_20210504-085625_config.pickle
-------------------------------









Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_1 (InputLaye

In [ ]:
import tensorflow.compat.v1 as tf
import os
os.environ['TF_CPP_MIN_LOG_LEVEL']='2'

In [ ]:
!python predict.py -i /content/drive/'MyDrive'/ObjectDetection-master/kaggle_simpson_testset -c ./save/train_20210503-110517_config.pickle

Using TensorFlow backend.










2021-05-04 08:56:20.060236: E tensorflow/stream_executor/cuda/cuda_driver.cc:318] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected






#[Problem 2] Code reading

1. RPN which is defined in resnet.py, 
```def rpn(base_layers,num_anchors)```
2. Roipooling which is defined in RoiPoolingConv.py,
```class RoiPoolingConv(Layer)```
3. Losses which are defined in losses.py,
```
def rpn_loss_regr(num_anchors)
def rpn_loss_cls(num_anchors)
def class_loss_regr(num_classes)
def class_loss_cls(y_true, y_pred)
```

#[Problem 3] Estimation by learned weights

In [ ]:
cd ..

/


In [ ]:
cd keras-yolo3

/content/drive/My Drive/keras-yolo3


In [ ]:
! python convert.py yolov3.cfg yolov3.weights model_data/yolo.h5

Using TensorFlow backend.
Loading weights.
Weights Header:  0 2 0 [32013312]
Parsing Darknet config.
Creating Keras model.


Parsing section net_0
Parsing section convolutional_0
conv2d bn leaky (3, 3, 3, 32)




2021-05-04 09:27:56.209150: E tensorflow/stream_executor/cuda/cuda_driver.cc:318] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected





Parsing section convolutional_1
conv2d bn leaky (3, 3, 32, 64)
Parsing section convolutional_2
conv2d bn leaky (1, 1, 64, 32)
Parsing section convolutional_3
conv2d bn leaky (3, 3, 32, 64)
Parsing section shortcut_0
Parsing section convolutional_4
conv2d bn leaky (3, 3, 64, 128)
Parsing section convolutional_5
conv2d bn leaky (1, 1, 128, 64)
Parsing section convolutional_6
conv2d bn leaky (3, 3, 64, 128)
Parsing section shortcut_1
Parsing section convolutional_7
conv2d bn leaky (1, 1, 128, 64)
Parsing section convolutional_8
conv2d bn leaky (3, 3, 64, 128)
Parsing section shortcut_2
Parsing section convolutional_

In [ ]:
!python yolo_video.py --image

Using TensorFlow backend.
Image detection mode
 Ignoring remaining command line arguments: ./path2your_video,



2021-05-04 09:47:48.038647: E tensorflow/stream_executor/cuda/cuda_driver.cc:318] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected









model_data/yolo.h5 model, anchors, and classes loaded.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
Input image filename:content/domestic-dog_thumb.jpg
Open Error! Try again!
Input image filename:../../../domestic-dog_thumb.jpg
(416, 416, 3)
Found 2 boxes for img
sofa 0.41 (116, 81) (2972, 2619)
dog 0.99 (202, 255) (2872, 2941)
4.788730917000066
Input image filename:Traceback (most recent call last):
  File "yolo_video.py", line 73, in <module>
  File "yolo_video.py", line 8, in detect_img
    img = input('Input image filename:')
KeyboardInterrupt


The results from a dog image:
- sofa 0.41 (116, 81) (2972, 2619)
- dog 0.99 (202, 255) (2872, 2941)

#[Problem 4] Create a file for learning

In [ ]:
import pandas as pd
annotation_df = pd.read_csv('../ObjectDetection-master/annotation.txt',header=None)
n_sample, n_col = annotation_df.shape
print(n_sample,n_col)
annotation_df.head()

7889 6


0  ...                       5
0  simpsons_dataset/abraham_grampa_simpson/pic_00...  ...  abraham_grampa_simpson
1  simpsons_dataset/abraham_grampa_simpson/pic_00...  ...  abraham_grampa_simpson
2  simpsons_dataset/abraham_grampa_simpson/pic_00...  ...  abraham_grampa_simpson
3  simpsons_dataset/abraham_grampa_simpson/pic_00...  ...  abraham_grampa_simpson
4  simpsons_dataset/abraham_grampa_simpson/pic_00...  ...  abraham_grampa_simpson

[5 rows x 6 columns]

In [ ]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
annotation_df.iloc[:,5] = le.fit_transform(annotation_df.iloc[:,5])
annotation_df.tail()

0    1   2    3    4   5
7884  simpsons_dataset/moe_szyslak/pic_0160.jpg   21   9   69   75  13
7885  simpsons_dataset/moe_szyslak/pic_0033.jpg   89   8  126   74  13
7886  simpsons_dataset/moe_szyslak/pic_0337.jpg   66  75  196  245  13
7887  simpsons_dataset/moe_szyslak/pic_1067.jpg  159  44  287  247  13
7888  simpsons_dataset/moe_szyslak/pic_0928.jpg  185  64  396  329  13

In [ ]:
path = '../ObjectDetection-master/annotation.txt'
out_path = 'train.txt'

with open(path) as f:
  lines = f.readline() 
  for i in range(n_sample):
    split_line = lines.split(',') 
    image_path = split_line[0]
    split_line[0] = './' + image_path 
    split_line[-1] = str(annotation_df.iloc[i,5]) + '\n'  

    with open(out_path, mode='a') as out_f:
      join_line = ','.join(split_line)  
      join_line = join_line.replace('.jpg,','.jpg ')  
      out_f.write(join_line)

#[Problem 5] Confirmation that learning can be done

In [ ]:
cp -R ../ObjectDetection-master/simpsons_dataset/ ./

^C


In [ ]:
!python convert.py -w yolov3.cfg yolov3.weights model_data/yolo_weights.h5

Using TensorFlow backend.
Loading weights.
Weights Header:  0 2 0 [32013312]
Parsing Darknet config.
Creating Keras model.


Parsing section net_0
Parsing section convolutional_0
conv2d bn leaky (3, 3, 3, 32)




2021-05-04 11:18:30.472729: E tensorflow/stream_executor/cuda/cuda_driver.cc:318] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected





Parsing section convolutional_1
conv2d bn leaky (3, 3, 32, 64)
Parsing section convolutional_2
conv2d bn leaky (1, 1, 64, 32)
Parsing section convolutional_3
conv2d bn leaky (3, 3, 32, 64)
Parsing section shortcut_0
Parsing section convolutional_4
conv2d bn leaky (3, 3, 64, 128)
Parsing section convolutional_5
conv2d bn leaky (1, 1, 128, 64)
Parsing section convolutional_6
conv2d bn leaky (3, 3, 64, 128)
Parsing section shortcut_1
Parsing section convolutional_7
conv2d bn leaky (1, 1, 128, 64)
Parsing section convolutional_8
conv2d bn leaky (3, 3, 64, 128)
Parsing section shortcut_2
Parsing section convolutional_

In [ ]:
!python train.py

Using TensorFlow backend.








2021-05-04 11:20:56.050306: E tensorflow/stream_executor/cuda/cuda_driver.cc:318] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected





Create YOLOv3 model with 9 anchors and 20 classes.
/usr/local/lib/python3.7/dist-packages/keras/engine/saving.py:1009: UserWarning: Skipping loading of weights for layer conv2d_59 due to mismatch in shape ((1, 1, 1024, 75) vs (255, 1024, 1, 1)).
  weight_values[i].shape))
/usr/local/lib/python3.7/dist-packages/keras/engine/saving.py:1009: UserWarning: Skipping loading of weights for layer conv2d_59 due to mismatch in shape ((75,) vs (255,)).
  weight_values[i].shape))
/usr/local/lib/python3.7/dist-packages/keras/engine/saving.py:1009: UserWarning: Skipping loading of weights for layer conv2d_67 due to mismatch in shape ((1, 1, 512, 75) vs (255, 512, 1, 1)).
  weight_values[i].shape))
/usr/local/lib/python3.7/dist-packages/keras/engine/saving.py:1009: UserWarning: Skipping loading of weigh

Even though the training is possible yet still it consumes lots of time, but from checking the losses, they are decreasing which means that its learning